doc/sphinx/language/core/records.rst

```diff
@@ -3,208 +3,313 @@
+
+.. table:: Printing Record @qualid
+
+   This :term:`table` specifies a set of qualids which are displayed as records.  Use the
+   :cmd:`Add` and :cmd:`Remove` commands to update the set of qualids.
 
-Syntax of Record Projections
+.. table:: Printing Constructor @qualid
 
-  .. insertprodn term_projection term_projection
+   This :term:`table` specifies a set of qualids which are displayed as constructors.  Use the
+   :cmd:`Add` and :cmd:`Remove` commands to update the set of qualids.
 
-  .. prodn::
-     term_projection ::= @term0 .( @qualid {? @univ_annot } {* @arg } )
-     | @term0 .( @ @qualid {? @univ_annot } {* @term1 } )
+.. flag:: Printing Projections
 
-The corresponding grammar rules are given in the preceding grammar. When :token:`qualid`
-denotes a projection, the syntax :n:`@term0.(@qualid)` is equivalent to :n:`@qualid @term0`,
-the syntax :n:`@term0.(@qualid {+ @arg })` to :n:`@qualid {+ @arg } @term0`.
-and the syntax :n:`@term0.(@@qualid {+ @term0 })` to :n:`@@qualid {+ @term0 } @term0`.
-In each case, :token:`term0` is the projected object and the
-other arguments are the parameters of the inductive type.
+   Activates the projection form (dot notation) for printing projections (off by default).
 
-Since the projected object is part of the notation,
-it is always considered an explicit argument of :token:`qualid`,
-even if it is formally declared as implicit (see :ref:`ImplicitArguments`),
+   .. example::
 
+      .. coqtop:: all
 
-.. note:: Records defined with the :cmd:`Record` command are not allowed to be
-   recursive (references to the record's name in the type of its field
-   raises an  error). To define recursive records, one can use the
-   :cmd:`Inductive` and :cmd:`CoInductive` commands, resulting in an inductive or coinductive record.
-   Definition of mutually inductive or coinductive records are also allowed, as long
-   as all of the types in the block are records.
+         Check top half.  (* off: application form *)
+         Set Printing Projections.
+         Check top half.  (* on:  projection form *)
 
-.. note:: Induction schemes are automatically generated for inductive records.
-   Automatic generation of induction schemes for non-recursive records
-   defined with the :cmd:`Record` command can be activated with the
-   :flag:`Nonrecursive Elimination Schemes` flag (see :ref:`proofschemes-induction-principles`).
+.. note:: Records exist in two flavors. In the first
+   implementation, a record :n:`@ident` with parameters :n:`{* @binder }`,
+   constructor :n:`@ident__0`, and fields :n:`{* @name @field_spec }`
+   is represented as a variant type with a single
+   constructor: :n:`Variant @ident {* @binder } : @sort := @ident__0
+   {* ( @name @field_spec ) }` and projections are defined by case analysis.
+   In the second implementation, records have
+   primitive projections: see :ref:`primitive_projections`.
 
 .. warn:: @ident cannot be defined.
```

Comment:
This warning seems should be split into several and presented closer to `ident cannot be defined because it is informative and ident is not`. Furthermore, at least the first point is inaccurate (this would be an error, not a warning) and the reference to `Axiom` is a bit unexpected.

The next two error messages should also be moved.

Reply:
I split off the first case.  It's not obvious how to map the other two cases to the messages generated by `Record.warning_or_error`, if indeed they match.  The third case may correspond to `MissingProj`, but I'm reluctant to describe it because I can't produce the message.  (We should have examples of how to produce every message even if we don't include all that in the doc.)

Reply:
Food for thought: what about having an output test suite file dedicated to triggering all the documented error messages. Then we could have a mechanism to alert reviewers when a PR updates this test suite file without updating the documentation (@coqbot could post a comment for instance).

Reply:
It would be good to have test suite files that generate as many of the messages as we can (some may be quite hard) just to document them for developers.  But how would we ensure that when a developer adds a new message that they remember to add it to the test suite?  Probably would need to invent a convention for tagging error message generation in the code so we can get a list of all error messages automatically.

Reply:
PS Shouldn't it be lint instead of coqbot?

Reply:
I believe it is possible to some extent already. Since we have errors in an open ended variant we could just print all the values. I'm sure people have done this before in OCaml. The issue remains that many errors in the codebase still rely on user_err etc. cc @ppedrot Who might be able to comment further.

Reply:
It would be hard to do currently because we use the standard `exn` type. Even if we somehow managed to trick OCaml into printing the whole set of exn constructors (e.g. loading a byte plugin and relying on the REPL) we would get a lot of unrelated stuff in there. I'd rather have some kind of explicit registering mechanism, which we kind of have but is not expressive enough to give you access to the actual exception.

Reply:
Folks: all your plans are fine, and I would definitely like some kind of registration mechanism for errors which can produce a dump of all known errors and can be checked against the documentation. But this is *not* what I was suggesting here. I was suggesting something more trivial (and therefore easier to implement): to adopt a *convention* that documented error messages are all listed in a specific test file, and to alert when this test file is modified but the documentation isn't (here I mentioned @coqbot rather than the linter because the alert would not be foolproof: it could happen that the test file needs to be updated but the doc doesn't, so it should not be blocking).

Reply:
That could be done, but might not be so trivial. I guarantee there are errors in the code that have no known way to triggering. At least with exceptions we can trigger them all manually and see what they output. I think maybe we could do some grep magic on `user_err` and related things to list and perhaps later lint these errors. However even that has it's flaws.

Reply:
>I guarantee there are errors in the code that have no known way to triggering.

If that's the case, shouldn't these errors be turned into anomalies?

In any case, my suggestion was not related to exhaustively listing all error messages in the documentation, but rather related to making sure that when error messages evolve or are removed, then the documentation of these error messages is kept up-to-date (we keep stumbling upon documented error messages that do not exist anymore). Do you also think that there are *documented* error messages that have no known way to triggering?